ch_tools/common/clickhouse/client/clickhouse_client.py

```diff
@@ -230,17 +238,36 @@ def query(
                 raise UserWarning(2, "Can't find any port in clickhouse-server config")
 
         logging.debug("Executing query: %s", query)
-        if found_port in [ClickhousePort.HTTPS, ClickhousePort.HTTP]:
-            return self._execute_http(
-                query,
-                format_,
-                post_data,
-                timeout,
-                stream,
-                per_query_settings,
-                found_port,
-            )
-        return self._execute_tcp(query, format_, found_port)
+
+        for attempt_count in range(0, S3_MAX_RETRY_COUNT):
+            try:
+                if found_port in [ClickhousePort.HTTPS, ClickhousePort.HTTP]:
+                    return self._execute_http(
+                        query,
+                        format_,
+                        post_data,
+                        timeout,
+                        stream,
+                        per_query_settings,
+                        found_port,
+                    )
+
+                return self._execute_tcp(query, format_, found_port)
+            except ClickhouseError as e:
```

Comment:
Handling of retries here somewhat clutters this function. I'd suggest moving retry logic in a separate class or function and pass it  `self._execute_http()` as a callable.